[FLINK-27159][table-api] Support first_value/last_value in the Table API

shuiqiangchen opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request added support for `first_value` and `last_value` in Table API.


## Brief change log

- Added `first_value` and `last_value` in BuiltInFunctionDefinitions
- Added `first_value` and `last_value` in SqlAggFunctionVisitor

## Verifying this change

This change has test case covered by testFirstLastValue() in AggregateITCase.scala

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

dianfu commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:
```diff
@@ -410,6 +410,20 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
                     .outputTypeStrategy(TypeStrategies.aggArg0(t -> t, true))
                     .build();
 
+    public static final BuiltInFunctionDefinition FIRST_VALUE =
```
Add last_value/first_value methods in BaseExpressions and also in PyFlink?

shuiqiangchen replied:
Ok, It was my omission

dianfu commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/AggregateITCase.scala:
```diff
@@ -114,6 +114,34 @@ class AggregateITCase(mode: StateBackendMode) extends StreamingWithStateTestBase
   }
 
+  @Test
+  def testFirstLastValue(): Unit = {
```
Could we integrate it into existing test case, e.g. testGroupAggregateWithExpression, etc. Then we could avoid introducing another it case.

dianfu reviewed (commented):
@shuiqiangchen Thanks for the PR. Have left a few minor comments. Besides, it would be great to also update the documentation: https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/functions/systemfunctions/.

dianfu commented:
@shuiqiangchen It seems that there is a new way provided to support built-in functions. See https://issues.apache.org/jira/browse/FLINK-20522 and https://github.com/apache/flink/commit/df46f8948279d1ed89d7113919b4c4f720cf4199 for more details.

shuiqiangchen commented:
@dianfu Thanks, I will refer to those commits in the ticket you mentioned.

dianfu commented:
@shuiqiangchen It seems that the new stack is used when adding new functions. It makes adding new functions easier. Regarding to this PR, its purpose is to expose existing functions first_value/last_value to Table API and so I guess we could still use the old stack.

shuiqiangchen commented:
@dianfu Yes, the implementation of LastValueAggFunction and FirstValueAggFunction have followed the new form. This PR is mainly to expose builtin functions to TableAPI.

shuiqiangchen commented:
@dianfu Thank you for your suggestions. I have updated the PR accordingly, please have a look.
Changes are as followed:
- added firstValue and lastValue in BaseExpressions
- added first_value and last_value in expression.py.
- Reused the existing ITCase.
- updated the documentation for System (Built-in) Functions

And there is a minor issue , I find that the functions documentation for python is redirect to System (Built-in) Functions documentation for Java API.  It's first_value() in python while firstValue() in Java, which seems unaligned that the doc might be a bit misleading.

dianfu commented:
> And there is a minor issue , I find that the functions documentation for python is redirect to System (Built-in) Functions documentation for Java API. It's first_value() in python while firstValue() in Java, which seems unaligned that the doc might be a bit misleading.
Good catch. We could find a way to improve this in a separate ticket.
